web-common/src/features/entity-management/file-artifact.ts

```diff
@@ -0,0 +1,297 @@
+  get,
+  type Readable,
+  writable,
+  type Writable,
+} from "svelte/store";
+import { runtimeServicePutFile } from "@rilldata/web-common/runtime-client";
+import { fileArtifacts } from "./file-artifacts";
+import {
+  DIRECTORIES_WITHOUT_AUTOSAVE,
+  FILES_WITHOUT_AUTOSAVE,
+} from "../editor/config";
+import { localStorageStore } from "@rilldata/web-common/lib/store-utils";
+import { parseKindAndNameFromFile } from "./file-content-utils";
+
+const UNSUPPORTED_EXTENSIONS = [".parquet", ".db", ".db.wal"];
+
+export class FileArtifact {
+  readonly path: string;
+  readonly name = writable<V1ResourceName | undefined>(undefined);
+  readonly reconciling = writable(false);
+  readonly localContent: Writable<string | null> = writable(null);
+  readonly remoteContent: Writable<string | null> = writable(null);
+  readonly hasUnsavedChanges = writable(false);
+  readonly fileExtension: string;
+  readonly ready: Promise<boolean>;
+  readonly fileTypeUnsupported: boolean;
+  readonly folderName: string;
+  readonly fileName: string;
+  readonly disableAutoSave: boolean;
+  readonly autoSave: Writable<boolean>;
+
+  private remoteCallbacks = new Set<(content: string) => void>();
+  private localCallbacks = new Set<(content: string | null) => void>();
+
+  // Last time the state of the resource `kind/name` was updated.
+  // This is updated in watch-resources and is used there to avoid
+  // unnecessary calls to GetResource API.
+  lastStateUpdatedOn: string | undefined;
+
+  constructor(filePath: string) {
+    const [folderName, fileName] = splitFolderAndName(filePath);
+
+    this.path = filePath;
+    this.folderName = folderName;
+    this.fileName = fileName;
+
+    this.disableAutoSave =
+      FILES_WITHOUT_AUTOSAVE.includes(filePath) ||
+      DIRECTORIES_WITHOUT_AUTOSAVE.includes(folderName);
+
+    if (this.disableAutoSave) {
+      this.autoSave = writable(false);
+    } else {
+      this.autoSave = localStorageStore<boolean>(`autoSave::${filePath}`, true);
+    }
+
+    this.fileExtension = extractFileExtension(filePath);
+    this.fileTypeUnsupported = UNSUPPORTED_EXTENSIONS.includes(
+      this.fileExtension,
+    );
+  }
+
+  updateRemoteContent = (content: string, alert = true) => {
+    this.remoteContent.set(content);
+    if (alert) {
+      for (const callback of this.remoteCallbacks) {
+        callback(content);
+      }
+    }
+  };
+
+  async fetchContent(invalidate = false) {
+    const instanceId = get(runtime).instanceId;
+    const queryParams = {
+      path: this.path,
+    };
+    const queryKey = getRuntimeServiceGetFileQueryKey(instanceId, queryParams);
+
+    if (invalidate) await queryClient.invalidateQueries(queryKey);
+
+    const queryFn: QueryFunction<
+      Awaited<ReturnType<typeof runtimeServiceGetFile>>
+    > = ({ signal }) => runtimeServiceGetFile(instanceId, queryParams, signal);
+
+    const { blob } = await queryClient.fetchQuery({
+      queryKey,
+      queryFn,
+      staleTime: Infinity,
+    });
```

Comment:
I believe this is correct. It's definitely still on my mind to figure out a way to make this cleaner, but it should be operating as intended.

`invalidateQueries` doesn't _really_ refetch anything. It marks queries as stale and, if they're being actively rendered using `createQuery`, then they'll be refetched and subscribers will be updated.

However, with this PR, there are no longer components that are subscribed to any file queries via `createRuntimeServiceGetFile`. This is mostly CodeMirror's "fault", whose API does not mesh well with the Svelte approach of reactive functions or derived queries.

If you try to comment out the invalidateQueries line or early return immediately after it, you'll see that the network request won't actually happen when you make a change in your IDE. You could set `{ refetchType: "all" }` on the invalidation and early return, which would fire the network request, but it wouldn't propagate to the editor because `updateRemoteContent`, which handles the callbacks and writable update, wasn't called.

Reply:
Ok, thanks for the detail! Makes sense that this is needed after removing component-native Query Observers.